Feat: add twitter share link to final tutorial resources page

zebateira opened the pull request:
This PR implements the share described in #243 

The link looks like so:
![image](https://user-images.githubusercontent.com/2353186/71994477-3cfad500-3252-11ea-8432-a0de216b7bbd.png)
(Let me know if you prefer the twitter share button instead with the twitter logo, since this simple implementation looks a bit bland).

Clicking on the link will open the following browser tab:
![image](https://user-images.githubusercontent.com/2353186/71994456-32d8d680-3252-11ea-9efa-3a38e4dded9c.png)

Let me know what is the final text you want to include in the tweet. We can add hashtags as well.

Closes #243 

Twitter reference: https://developer.twitter.com/en/docs/twitter-for-websites/tweet-button/guides/parameter-reference1

zebateira commented:
> Either that or replacing Share with something more meaningful and specific to twitter, like Share on Twitter or Tweet, so the user knows what kind of sharing that button is for.

Agreed - aesthetically I'd go for the change of text, but it might go unnoticed by the user (as it currently is as well).

what do you think @terichadbourne ?

terichadbourne commented:
Yeah, I agree that we need more context in the message. It would be great
to have a section appear (if all lessons are passed) that says something
like "Congratulations on completing the Tutorial Name tutorial!" (maybe
incorporating our checkmark icon?) and then something like "Share your
success on Twitter" as the button text. It probably should not be right
next to the More Tutorials button. Maybe it's a full-width box on a shaded
background above or below the resource listings?

Within the content of the tweet it would be great to include the tutorial
name (full name, not shortname) unless we're hitting the character limit.
Could we snag the relevant "project" from the tutorials.json file (ipfs,
libp2p) and add that as a hashtag as well?

On Wed, Jan 8, 2020 at 11:53 AM Zé Bateira <notifications@github.com> wrote:

> Either that or replacing Share with something more meaningful and specific
> to twitter, like Share on Twitter or Tweet, so the user knows what kind of
> sharing that button is for.
>
> Agreed - aesthetically I'd go for the change of text, but it might go
> unnoticed by the user (as it currently is as well).
>
> what do you think @terichadbourne <https://github.com/terichadbourne> ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProtoSchool/protoschool.github.io/pull/340?email_source=notifications&email_token=AESIRCOIN7PBTU7FB2PKHCDQ4YAHBA5CNFSM4KEKWS42YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEINHGTI#issuecomment-572158797>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AESIRCKYWXHOWWIVT4X5WDLQ4YAHBANCNFSM4KEKWS4Q>
> .
>


-- 
*Teri Chadbourne, CMP*
Community Manager & Lead Maintainer, ProtoSchool <https://proto.school>
Protocol Labs <https://protocol.ai/>

Twitter <https://twitter.com/eventteri> | GitHub
<https://github.com/terichadbourne>

zebateira commented:
> Yeah, I agree that we need more context in the message. It would be great
to have a section appear (if all lessons are passed) that says something
like "Congratulations on completing the Tutorial Name tutorial!" (maybe
incorporating our checkmark icon?) and then something like "Share your
success on Twitter" as the button text. It probably should not be right
next to the More Tutorials button. Maybe it's a full-width box on a shaded
background above or below the resource listings?

Fully agree - I was sad to not have a congratulatory message when I was playing around with the tutorials ahah.
I think it would make more sense to put it above the resource listings.

> Within the content of the tweet it would be great to include the tutorial
name (full name, not shortname) unless we're hitting the character limit.
Could we snag the relevant "project" from the tutorials.json file (ipfs,
libp2p) and add that as a hashtag as well?

So, add an hashtag with the name of the tutorial?

terichadbourne commented:
> So, add an hashtag with the name of the tutorial?

No, add the *name* of the tutorial to the text of the tweet but add a
hashtag for the *project* (#IPFS, #libp2p, etc) - here's an example
<https://github.com/ProtoSchool/protoschool.github.io/blob/58397b1a10a158a89b18f1068c3b7f4ddfe41332/src/static/tutorials.json#L4>
of
where the project is listed in `tutorials.json`. The project field makes
the correct logo appear in the cards/listings for each tutorial. All of the
current content happens to be about IPFS but we'll need to future-proof the
code to pull from that field.


w/r/t wanting more congratulations when you complete a tutorial, one option
is to create the little display you're building as a component and have it
appear both on the resources page and at the bottom of each lesson (or just
the last lesson in a tutorial). The last lesson in a tutorial is the most
common - but not only - place from which someone would flip over into fully
completed status, and if we used a colored background (maybe the aqua from
the top nav?) it could be a nice callout. That way you'd see it even if you
didn't choose to visit the resources page.

On Wed, Jan 8, 2020 at 12:20 PM Zé Bateira <notifications@github.com> wrote:

> Yeah, I agree that we need more context in the message. It would be great
> to have a section appear (if all lessons are passed) that says something
> like "Congratulations on completing the Tutorial Name tutorial!" (maybe
> incorporating our checkmark icon?) and then something like "Share your
> success on Twitter" as the button text. It probably should not be right
> next to the More Tutorials button. Maybe it's a full-width box on a shaded
> background above or below the resource listings?
>
> Fully agree - I was sad to not have a congratulatory message when I was
> playing around with the tutorials ahah.
> I think it would make more sense to put it above the resource listings.
>
> Within the content of the tweet it would be great to include the tutorial
> name (full name, not shortname) unless we're hitting the character limit.
> Could we snag the relevant "project" from the tutorials.json file (ipfs,
> libp2p) and add that as a hashtag as well?
>
> So, add an hashtag with the name of the tutorial?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProtoSchool/protoschool.github.io/pull/340?email_source=notifications&email_token=AESIRCMDBPRG5YNY3WKOTMTQ4YDONA5CNFSM4KEKWS42YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEINJ76Q#issuecomment-572170234>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AESIRCOA5MJFNDNQMHSIEW3Q4YDONANCNFSM4KEKWS4Q>
> .
>


-- 
*Teri Chadbourne, CMP*
Community Manager & Lead Maintainer, ProtoSchool <https://proto.school>
Protocol Labs <https://protocol.ai/>

Twitter <https://twitter.com/eventteri> | GitHub
<https://github.com/terichadbourne>

zebateira commented:
> That way you'd see it even if you
didn't choose to visit the resources page.

I'd argue that most users will go to the resources page since the button says "Next" which still invites the user to "complete" the tutorial. I'll add it to the Resources page and then you can try it out and see if you still prefer to show the callout to the final lesson 👍

zebateira commented:
Ok, so I removed the link on the bottom and added this to the top of the resources page:
![image](https://user-images.githubusercontent.com/2353186/72080958-105cc100-3317-11ea-9a8d-870e4ca092f9.png)

It only shows if the full tutorial has been completed.

And this is how the new tweet text looks:
![image](https://user-images.githubusercontent.com/2353186/72081169-6f223a80-3317-11ea-9156-d9f76faf0ae8.png)

dominguesgm commented on src/utils/tutorials.js:
```diff
@@ -0,0 +1,10 @@
+import tutorialsList from '../static/tutorials.json'
+
+export const getCurrentTutorial = route => (
+  route.props && route.props.default && route.props.default.tutorialId &&
```
If vue-cli 3 already supported optional chaining we could have this a bit prettier 😄  but i'm pretty sure it doesn't.

zebateira replied:
Yeah we would need to add the babel plugin manually.
Another solution would be to add `lodash` and use `lodash.get`. Since it is already in the dependency tree, no bundle size issues there.

terichadbourne commented on src/components/CongratulationsCallout.vue:
```diff
@@ -0,0 +1,48 @@
+  },
+  data () {
+    let href = 'https://twitter.com/intent/tweet?'
+    href += `text=I just completed the tutorial "${this.tutorial.title}" at @ProtoSchool!`
```
```suggestion
    href += `text=I just completed the ${this.tutorial.title} tutorial at @ProtoSchool!`
```

terichadbourne commented on src/components/CongratulationsCallout.vue:
```diff
@@ -0,0 +1,48 @@
+    <img src="../static/images/complete.svg" class="w2 h2 mr3" alt="complete" />
+    <div class="flex items-start flex-column">
+      <div class="lh-title">
+        Congratulations on completing the <span class="b">"{{tutorial.title}}"</span> tutorial!
```
```suggestion
        Congratulations on completing the <span class="b">{{tutorial.title}}</span> tutorial!
```

terichadbourne commented on src/components/Tutorial.vue:
```diff
@@ -50,11 +51,11 @@ export default {
     LessonLink
   },
-  data: self => {
+  data () {
```
Could you help me understand why this one doesn't need the `self` and others instances do?

zebateira replied:
in short, by using [`method definitions`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Method_definitions) we can use `this.*` inside the `data` function.

but actually I'll revert this to be consistent with the current code base - although I think it's better to use `this.*`, `self` prop works just as fine.

terichadbourne commented on src/main.js:
```diff
@@ -70,52 +70,52 @@ const routes = [
-  { path: '/data-structures/03', component: T0001L03 },
-  { path: '/data-structures/04', component: T0001L04 },
-  { path: '/data-structures/05', component: T0001L05 },
+  { path: '/data-structures/01', component: T0001L01, props: { tutorialId: '0001' } },
```
These changes are to a part of the code created by the tutorial author, so we'll have to update our documentation. Can you please update the Developing Tutorials doc accordingly as part of this PR?

zebateira replied:
done

terichadbourne commented on src/components/Tutorial.vue:
```diff
@@ -50,11 +51,11 @@ export default {
     LessonLink
   },
-  data: self => {
+  data () {
```
Could you help me understand why this instance of `data` doesn't need the `self` but others do?

terichadbourne commented:
This is looking great @zebateira. Added a few suggestions/questions for you. 

> I'd argue that most users will go to the resources page since the button says "Next" which still invites the user to "complete" the tutorial. I'll add it to the Resources page and then you can try it out and see if you still prefer to show the callout to the final lesson

You're totally right. In my head the button said "More to Explore." 😂

zebateira commented:
@terichadbourne let me know what you think 🙏

zebateira commented:
> Added a few suggestions/questions for you.

Approved 🙌

zebateira commented:
conflicts solved and rebased as well. ready to go @terichadbourne @dominguesgm  🚀

terichadbourne commented:
Looks great! I fixed the caps in tutorials.json because it was hard to parse this in the tweet (and looks like a capitalization accident: 

> I just completed the Decentralized data structures tutorial at @ProtoSchool! http://localhost:3000/data-structures #IPFS

versus this (new version): 

> I just completed the Decentralized Data Structures tutorial at @ProtoSchool! http://localhost:3000/data-structures #IPFS

terichadbourne commented:
@zebateira Just noticed my Tweet came out with localhost in the URL. Do you think we should edit `export const getTutorialFullUrl = tutorial => `${window.location.origin}/${tutorial.url}`` to force `https://proto.school`, or no point bothering?

zebateira commented:
@terichadbourne I just did it like so to avoid hardcoding it

terichadbourne commented:
I assume it's fine as long as no one accidentally sends a tweet while
testing. Feel free to merge if you prefer as is.

On Fri, Jan 10, 2020 at 11:24 AM Zé Bateira <notifications@github.com>
wrote:

> @terichadbourne <https://github.com/terichadbourne> I just did it like so
> to avoid hardcoding it
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProtoSchool/protoschool.github.io/pull/340?email_source=notifications&email_token=AESIRCL3OLZICZEHHHIXYNLQ5COLFA5CNFSM4KEKWS42YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIUOA2I#issuecomment-573104233>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AESIRCJO7E673UTV3BO3AUDQ5COLFANCNFSM4KEKWS4Q>
> .
>


-- 
*Teri Chadbourne, CMP*
Community Manager & Lead Maintainer, ProtoSchool <https://proto.school>
Protocol Labs <https://protocol.ai/>

Twitter <https://twitter.com/eventteri> | GitHub
<https://github.com/terichadbourne>
